### PR TITLE
run_sorter in docker/singlurity : add option for installing spikeinterface

### PR DESCRIPTION
**Pull request by samuelgarcia**
This depend on #2271

* add a global DEV_MODE to replace the dev in versioning to check if release or not
* add installation_mode in run_sorter_container()

@alejoe91 tell if this is what we said
@alejoe91 : can you help to test ?

### CODE REVIEW

File: src/spikeinterface/sorters/runsorter.py [hunk 1]
```diff
@@ -463,15 +497,30 @@ def run_sorter_container(
...
+    if installation_mode == "auto":
+        if DEV_MODE:
+            if is_editable_mode():
+                installation_mode = "dev"
+            else:
+                installation_mode = "github"
+        else:
+            installation_mode = "github"
+        if verbose:
+            print(f"installation_mode='auto' is switch to '{installation_mode}'")
```
Comment (zm711) [inline src/spikeinterface/sorters/runsorter.py:510]:
```suggestion
            print(f"installation_mode='auto' switching to installation_mode: '{installation_mode}'")
```

File: src/spikeinterface/sorters/runsorter.py [hunk 2]
```diff
@@ -463,15 +497,30 @@ def run_sorter_container(
...
+
+    if installation_mode == "folder":
+        assert spikeinterface_folder_source is not None, "for installation_mode='folder', spikeinterface_folder_source must provided"
```
Comment (zm711) [inline src/spikeinterface/sorters/runsorter.py:513]:
```suggestion
        assert spikeinterface_folder_source is not None, "for installation_mode='folder', spikeinterface_folder_source must be provided"
```

File: src/spikeinterface/sorters/runsorter.py [hunk 3]
```diff
@@ -414,9 +436,15 @@ def run_sorter_container(
...
-        json.dumps(check_json(rec_dict), indent=4), encoding="utf8"
-    )
+    if recording.check_serializability("json"):
+        (parent_folder / "in_container_recording.json").write_text(
+            json.dumps(check_json(rec_dict), indent=4), encoding="utf8"
+        )
+    elif recording.check_serializability("pickle"):
+        (parent_folder / "in_container_recording.pickle").write_bytes(pickle.dumps(rec_dict))
+    else:
+        raise RuntimeError("To use run_sorter with container the recording must serializable")
```
Comment (zm711) [inline src/spikeinterface/sorters/runsorter.py:446]:
```suggestion
        raise RuntimeError("To use run_sorter with a container the recording must be serializable")
```
Sam spelled serializable perfectly here!

File: src/spikeinterface/__init__.py [hunk 4]
```diff
@@ -25,3 +25,9 @@
 import spikeinterface.full as si
 
 """
+
+# This flag must set to False for release
+# This avoid to use versionning that contains ".dev0" (and this is a better choice)
+# This mainly usefull when using run_sorter in container and spikeinterface install
```
Comment (zm711) [inline src/spikeinterface/__init__.py:31]:
```suggestion
# This flag must be set to False for release
# This avoids using versioning that contains ".dev0" (and this is a better choice)
# This is mainly useful when using run_sorter in a container and spikeinterface install
```
not quite sure what the very end of this comment means

**Review by zm711 — COMMENTED** [review]
Just docstring comments

File: src/spikeinterface/sorters/runsorter.py [hunk 5]
```diff
@@ -500,58 +549,69 @@ def run_sorter_container(
...
+            if DEV_MODE:
+                # not released yet use main branch
+                cmd = "pip install --user --upgrade --no-input --no-build-isolation git+https://github.com/SpikeInterface/spikeinterface.git@main#egg=spikeinterface[full]"
                 res_output = container_client.run_command(cmd)
-                cmd = f"pip install --user {si_source_folder}/spikeinterface[full]"
             else:
-                si_source = "remote repository"
-                cmd = "pip install --user --upgrade --no-input git+https://github.com/SpikeInterface/spikeinterface.git#egg=spikeinterface[full]"
+                # already released and has a tag 
+                si_version = '0.99.1'
```
Comment (alejoe91) [inline src/spikeinterface/sorters/runsorter.py:592]:
```suggestion
```
Can't we just use the package version?

File: src/spikeinterface/sorters/runsorter.py [hunk 6]
```diff
@@ -500,58 +549,69 @@ def run_sorter_container(
...
+            # this is "dev" + "folder"
             if verbose:
-                print(f"Installing dev spikeinterface from {si_source}")
+                print(f"Installing spikeinterface from local sources in {container_image}")
+
+            # copy spikeinterface sources to avoid problems between host and containers
+            cmd = f"mkdir {container_folder_source}"
             res_output = container_client.run_command(cmd)
-            cmd = "pip install --user --upgrade --no-input https://github.com/NeuralEnsemble/python-neo/archive/master.zip"
+            cmd = f"cp -r {container_folder_source_ro} {container_folder_source}"
```
Comment (alejoe91) [inline src/spikeinterface/sorters/runsorter.py:604]: Are you sure you don't need a `cmd = f"pip install --user {container_folder_source}/spikeinterface[full]"` here?

File: src/spikeinterface/sorters/runsorter.py [hunk 7]
```diff
@@ -500,58 +549,69 @@ def run_sorter_container(
...
+            cmd = f"pip install --user --no-input {container_folder_source}/spikeinterface[full]"
             res_output = container_client.run_command(cmd)
-    else:
-        # TODO version checking
-        if verbose:
-            print(f"spikeinterface is already installed in {container_image}")
+
+        if installation_mode == "dev":
+            # also install neo from github
+            cmd = "pip install --user --upgrade --no-input https://github.com/NeuralEnsemble/python-neo/archive/master.zip"
```
Comment (alejoe91) [inline src/spikeinterface/sorters/runsorter.py:612]: Should we have the same mechanism for NEO and probeinterface at this point?

File: src/spikeinterface/sorters/runsorter.py (hunk 5 above)
Comment (samuelgarcia) [reply src/spikeinterface/sorters/runsorter.py:592]: it was for testing I forgot to remove

File: src/spikeinterface/sorters/external/tests/test_singularity_containers.py [hunk 8]
```diff
@@ -35,6 +35,8 @@ def generate_run_kwargs():
     test_recording.set_channel_offsets(0)
     run_kwargs = dict(recording=test_recording, verbose=True)
     run_kwargs["singularity_image"] = True
+    run_kwargs["installation_mode"] = "folder"
+    run_kwargs["spikeinterface_folder_source"] = os.getenv("SPIKEINTERFACE_DEV_PATH")
```
[the remark this answers is not in the document]
Comment (samuelgarcia) [reply src/spikeinterface/sorters/external/tests/test_singularity_containers.py:39]: I think no because we do not install in editable mode in CI

**Comment by samuelgarcia** [comment]
@alejoe91 : I did a function for installing packages inside the container.
Can you have a look ?
I did local test with tridesclous with folder+pypi+gihub
It works : can you make some more tests ?

File: src/spikeinterface/sorters/runsorter.py [hunk 9]
```diff
@@ -513,68 +397,69 @@ def run_sorter_container(
         py_user_base_folder = parent_folder / "in_container_python_base"
         py_user_base_folder.mkdir(parents=True, exist_ok=True)
         py_user_base_unix = path_to_unix(py_user_base_folder)
-        si_source_folder = f"{py_user_base_unix}/sources"
+        container_folder_source = f"{py_user_base_unix}/sources"
```
Comment (samuelgarcia) [inline src/spikeinterface/sorters/runsorter.py:400]: Remove this

File: src/spikeinterface/sorters/runsorter.py [hunk 10]
```diff
@@ -513,68 +397,69 @@ def run_sorter_container(
...
     else:
-        si_source_folder = "/sources"
+        container_folder_source = "/sources"
```
Comment (samuelgarcia) [inline src/spikeinterface/sorters/runsorter.py:402]: Remove this

**Comment by samuelgarcia** [comment]
thank you camarade.
this is OK for me.

File: src/spikeinterface/sorters/external/tests/test_singularity_containers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alejoe91) [reply src/spikeinterface/sorters/external/tests/test_singularity_containers.py:39]: We do :) https://github.com/SpikeInterface/spikeinterface/blob/main/.github/workflows/test_containers_singularity.yml#L35

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**